### PR TITLE
fix(rootly,clerk,jira): reject path-traversal ids in URL segments

### PR DESCRIPTION
**Pull request by waleedlatif1**
## The defect

Rootly, Clerk and Jira tools interpolated LLM-writable identifiers (incident id, alert id, user id, organization id, session id, issue key, comment id, worklog id, attachment id, project id, …) straight into request path segments. Those params are `visibility: 'user-or-llm'`, so prompt injection controls them.

A value like `../../users/victim` escapes its API prefix once `fetch` normalizes the URL, re-aiming the request — carrying the workspace's credential — at a different resource on the same host, including on DELETE. `assertRequestUrlMatchesTrust` in `tools/request-transport.ts` only canonicalizes internal `/api/` routes, so nothing downstream catches it.

Clerk is the worst case: its credential is a backend API key with full user-management authority, and the affected routes include delete/ban/lock.

`encodeURIComponent` does not close this:

```js
encodeURIComponent('..')                                          // => '..'
new URL('https://api.clerk.com/v1/users/../ban').pathname         // => '/v1/ban'
new URL('https://api.clerk.com/v1/users/%2e%2e/ban').pathname     // => '/v1/ban'
```

`.` and `..` are unreserved, so they survive encoding verbatim, and the WHATWG URL parser removes dot segments *after* decoding. Only rejecting the value works.

## The fix

`safeUrlPathSegment(value, paramName)` from `@/tools/url-path` at **69 path-segment sites across 66 tool files**:

| service | sites | files |
|---|---|---|
| Rootly | 23 | 23 |
| Clerk | 26 | 23 |
| Jira | 20 request builders + 20 `transformResponse` fallbacks | 20 |

Jira tools build the URL twice — once in `request.url` when `cloudId` is already present, and again inside `transformResponse` after resolving the `cloudId`. Both were unguarded; both are guarded now.

Clerk's `provider` slug is guarded too: it is only *prefixed* with `oauth_`, so a `/` inside it still traverses (`provider = "x/../../users/victim"`).

No param `visibility`, no subBlock id, and no behaviour for legitimate input changed. `bun run tool-metadata:generate` produces no diff.

## Jira's `cloudId` — left unguarded on purpose

`cloudId` reaches a path segment on every Jira tool, but it is not LLM-writable:

- declared `visibility: 'hidden'` on every tool
- `grep -c cloudId blocks/blocks/jira.ts` is `0`, and `git log --all -S'cloudId' -- blocks/blocks/jira.ts` returns no commits — no revision has ever been able to set it
- its only real source is `getJiraCloudId` → `resolveAtlassianCloudId`, which returns a UUID from Atlassian's own accessible-resources endpoint

Guarding it would assert a threat model that does not exist. The test pins it instead, and says why.

Also checked and left alone: `lib/internal/jira/operations.ts`, the server side of `jira_update` / `jira_write` / `jira_add_attachment`, already runs `validateJiraIssueKey` (a `validatePathSegment` with `allowDots: false`) before every `client.issuePath()` interpolation.

## Tests

`path_safety.test.ts` under each of the three tool folders, modelled on `tools/vercel/edge_config_path_safety.test.ts`:

- **enumerates the tool barrel**, then per tool every parameter that reaches a path segment — so a new tool, or a new id param on an existing tool, is covered without registration
- **fuzzes one parameter at a time** while its siblings hold a safe value. Fuzzing them all together let a still-guarded sibling throw first and mask an unguarded one — that is a real hole this design closes
- resolves every URL with `new URL(...)`, the same normalization `fetch` performs, never string matching
- keeps the bare `.` and `..` vectors, plus `'  ..  '`, `%2f` spellings, backslashes, embedded traversal, and query/fragment injection
- a `LEGITIMATE_IDS` list asserts real values pass through byte-for-byte: `PROJ-123`, `MY_PROJECT-4567`, `user_2abcDEF`, `org_2abcDEF`, `sess_2abcDEF`, Rootly UUIDs, and the `..foo` / `foo..` edge cases

1684 tests pass across the three folders.

**Verified the tests can fail**: reverted the guard on `rootly_delete_incident.incidentId`, `clerk_delete_user.userId`, and `jira_delete_comment.commentId` one at a time and watched exactly that pair go red (10 failures each — the traversal vectors, both bare-dot rejections, and query injection), then restored.

## Gates

`bun run lint`, `bun run check:audits` (39 audits green, including `check:tool-request-boundary` and `check:api-validation:strict`), `check-block-registry.ts origin/staging`, `bun run type-check` clean on all touched files, `tool-metadata:generate` no-diff.

### CODE REVIEW

File: apps/sim/tools/jira/path_safety.test.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+function segmentsOf(tool: AnyTool, paramName: string, value: string): string[] {
+  return buildUrl(tool, paramName, value).pathname.split('/')
+}
+
+/** Every (tool, parameter) pair whose value lands in a URL path segment. */
+const PATH_PARAMS = Object.values(jiraTools)
+  .filter(isJiraTool)
+  .filter((tool) => typeof tool.request?.url === 'function')
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/jira/path_safety.test.ts]: Valid, and thank you — this is the same family of blindness as the bug the PR fixes, one level up.

You are exactly right about the mechanism: a Jira tool invoked without a `cloudId` sends its *configured* request to the fixed accessible-resources endpoint and builds the real per-site URL inside `transformResponse` after resolving one. `PATH_PARAMS` only inspects what `request.url` returns, so `jira_bulk_read.projectId` — which reaches a path segment **only** through that second construction — was filtered out and never fuzzed. The guard was there; the test could not see it.

Fixed generically rather than by hard-coding the one known case. `29f4547` adds a second suite that runs `transformResponse` with the `cloudId` withheld against a stubbed `fetch`, and reads back the URLs the tool actually asked for. Discovery is probe-driven exactly like `PATH_PARAMS`, so a future tool or a new ID param is covered without registration.

It finds **24** (tool, param) pairs — a strict superset of the 23 `request.url` exposes, the extra one being `jira_bulk_read :: projectId`. There is an explicit assertion pinning that, so the coverage cannot silently regress:

```ts
it('reaches jira_bulk_read.projectId, which request.url cannot expose', () => {
  expect(FALLBACK_PATH_PARAMS.map((entry) => entry.label)).toContain('jira_bulk_read :: projectId')
})
```

Verified it can fail: reverting that guard to the pre-PR `encodeURIComponent` turns exactly the `..` and `'  ..  '` vectors on `jira_bulk_read :: projectId` red. Extending discovery surfaced **no additional unguarded parameter** — every other fallback pair was already covered by the first suite.

Two notes on the design. Assertions in the new block are shape-based (same segment count, same non-probe segments) rather than whole-sequence equality, because a tool may legitimately issue a different number of requests for different inputs — and segment count is precisely what a popped dot segment changes. And the file header now states the residual limit honestly: this discovers URLs a tool *sends*, so a URL built but never fetched would still be invisible to both mechanisms. Rootly and Clerk headers record why a single suite is total there (no tool in either service fetches from `transformResponse`).

File: apps/sim/tools/clerk/path_safety.test.ts [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  'v1.2.3',
+  '..foo',
+  'foo..',
+] as const
+
+const SAFE_ID = 'SAFEID'
+
+/** Sentinel for the one parameter under test, so its slots are identifiable. */
+const PROBE_ID = 'PROBEID'
+
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/clerk/path_safety.test.ts:78]: Valid — CLAUDE.md forbids `any`, and the harness had it in both places you name (`ToolConfig<any, any>` and the `as any` at the URL-builder call site). Fixed in `29f4547` across all three files; no `any` remains in any of them.

Replaced with an `unknown` -> narrow type guard returning only the slice the suite actually drives:

```ts
type ToolParams = Record<string, unknown>
type UrlBuilder = (params: ToolParams) => string

interface PathBuildingTool {
  readonly id: string
  readonly params: Readonly<Record<string, ParamDefinition>>
  readonly buildUrl: UrlBuilder
  /** Present on the Jira tools that re-issue the call after resolving a cloudId. */
  readonly transformResponse: TransformResponse | null
}

function asPathBuildingTool(value: unknown): PathBuildingTool | null { /* ... */ }
```

Discovery becomes `.map(asPathBuildingTool).filter((tool): tool is PathBuildingTool => tool !== null)`, which also collapses the old two-step "is it a tool / does it have a URL builder" filter into one narrowing pass.

One deliberate choice worth flagging: the harness stays agnostic about each tool's own param and response generics rather than naming a concrete `ToolConfig<P, R>`. Those generics differ per tool and the suite iterates a heterogeneous barrel, so any single instantiation would be a fiction — the honest contract is the structural one above, and it is what the suite genuinely relies on. Everything downstream (`buildParams`, `buildUrl`, `segmentsOf`) is now typed end to end with no casts at the call sites.

Same finding was raised on #7262 against the GitHub path-safety harness; this is the shape I would apply there too, so the two read as siblings.

**Comment by waleedlatif1** [comment]
Both review threads addressed in `29f4547` and resolved.

**cubic P2 — `transformResponse`-built URLs were invisible to discovery.** Correct, and the same family of blindness as the bug this PR fixes. `PATH_PARAMS` only inspected `request.url`, so `jira_bulk_read.projectId` — which reaches a path segment only through the URL built inside `transformResponse` after `cloudId` resolution — was never fuzzed. Fixed generically: a second suite runs `transformResponse` with the `cloudId` withheld against a stubbed `fetch` and reads back the URLs requested, discovering parameters by probe exactly as the first suite does. It finds 24 (tool, param) pairs, a strict superset of the 23 `request.url` exposes. Reverting that guard to the pre-PR `encodeURIComponent` now fails. No additional unguarded parameter surfaced.

**Greptile P2 — harness erased tool types.** Correct; CLAUDE.md forbids `any`. `ToolConfig<any, any>` and the `as any` call-site cast are gone from all three files, replaced by `asPathBuildingTool`, an `unknown` -> narrow type guard returning the structural slice the suite drives.

On `cloudId`, restated for the record since it is the one thing deliberately left unguarded: it is `visibility: 'hidden'` on every Jira tool, `grep -c cloudId blocks/blocks/jira.ts` is `0` and `git log --all -S'cloudId' -- blocks/blocks/jira.ts` returns no commits (no revision has ever been able to set it), and its only real source is `resolveAtlassianCloudId`, which returns a UUID from Atlassian's own accessible-resources endpoint. Guarding it would assert a threat model that does not exist; the tests pin it and the header says why.

Gates re-run on the new commit: 2094 tests pass across the three folders, `bun run lint`, `bun run check:audits` (39 green), `check-block-registry.ts origin/staging`, `type-check` clean on all touched files, `tool-metadata:generate` no-diff.

@greptile

**Comment by waleedlatif1** [comment]
@greptile review

Both threads are addressed and resolved as of `29f4547`; your follow-up confirms no further concerns. Requesting a fresh full review so the summary reflects the current head rather than the original `8f981fa` — the score on this PR is still the one from before the typed harness and the Jira `transformResponse` coverage landed.

**Comment by waleedlatif1** [comment]
Closing for now — not because of a defect. This batch grew to 17 PRs across ~700 changed call sites, and we would rather revisit it as smaller, independently testable pieces than merge this much at once.

Nothing here is lost: the branch `fix/rootly-clerk-jira-path-safety` is preserved and this PR can be reopened. Review state, the reasoning on every thread, and the red-first verification all stay attached.
